### PR TITLE
Add support for multiprocess builds ("/m") in .NET Core MSBuild

### LINKED ISSUE #303: Enable /m in CoreCLR (xplat)
This features was turned off due to named pipe limitations and binary serialization missing from .NET Core. This feature should be enabled and supported.

### PR DESCRIPTION
**Pull request by dsplaisted**
Resolves #303

### CODE REVIEW

File: src/XMakeBuildEngine/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs [hunk 1]
```diff
@@ -390,12 +393,19 @@ private int LaunchNode(string msbuildLocation, string commandLineArgs)
 
             BackendNativeMethods.PROCESS_INFORMATION processInfo = new BackendNativeMethods.PROCESS_INFORMATION();
 
-            string appName = msbuildLocation;
+
+#if RUNTIME_TYPE_NETCORE
+            string coreRunName = NativeMethodsShared.IsUnixLike ? "CoreRun" : "CoreRun.exe";
```
Comment (akoeplinger) [inline src/XMakeBuildEngine/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs]: I think this should be `corerun` on non-Windows (at least according to https://github.com/dotnet/buildtools/blob/c578e891af4dea18b8f7e2524780bd4a9a3f8ce4/src/Microsoft.DotNet.Build.Tasks/PackageFiles/toolruntime.targets#L12)

**Comment by AndyGerlicher** [comment]
LTGM. Do we need to update the handshake so there's no possibility of mixing nodes?

Also, perhaps we should emit a warning that this is experimental use at your own risk until we're comfortable with the security implications?

And we should make this the default in our self-host scenario.

**Comment by dsplaisted** [comment]
OK, I've switch to use a pair of anonymous pipes on .NET Core instead of a single duplex named pipe, which makes multiprocess builds actually work on Linux and also resolves the security implications.  Enough has changed that I think it's worth re-reviewing the whole thing.

File: src/Shared/NodeEndpointOutOfProcBase.cs [hunk 2]
```diff
@@ -312,7 +345,15 @@ private void InitializeAsyncPacketThread()
...
         private void PacketPumpProc()
         {
+#if FEATURE_NAMED_PIPES_FULL_DUPLEX
             NamedPipeServerStream localPipeServer = _pipeServer;
+            PipeStream localWritePipe = _pipeServer;
+            PipeStream localReadPipe = _pipeServer;
+#else
+            PipeStream localWritePipe = _pipeClientToServer;
+            PipeStream localReadPipe = _pipeServerToClient;
+#endif
```
Comment (rainersigwald) [inline src/Shared/NodeEndpointOutOfProcBase.cs]: Nice way to hide this difference later :+1:

File: src/Shared/NodePacketTranslator.cs [hunk 3]
```diff
@@ -341,7 +341,30 @@ public void TranslateDotNet<T>(ref T value)
                 BinaryFormatter formatter = new BinaryFormatter();
                 value = (T)formatter.Deserialize(_packetStream);
             }
+#else
+            public void TranslateBuildEventArgs(ref BuildEventArgs value)
+            {
+                var serializedBuildEventArgs = new SerializedBuildEventArgs();
+                serializedBuildEventArgs.Translate(this);
+                value = SerializedBuildEventArgs.ToEventArgs(serializedBuildEventArgs);
```
Comment (rainersigwald) [inline src/Shared/NodePacketTranslator.cs]: What's the logic of mutating the input?

File: src/Shared/NodePacketTranslator.cs [hunk 4]
```diff
@@ -1027,5 +1070,407 @@ public bool TranslateNullable<T>(T value)
...
+                translator.Translate(ref fieldName);
+
+                if (translator.Mode == TranslationDirection.ReadFromStream)
+                {
+                    LoadedType loadedDclaringType = _typeLoader.Load(declaringTypeFullName, assemblyLoadInfo);
+                    Type declaringType = loadedDclaringType.Type;
+                    Field = declaringType.GetField(fieldName, BindingFlags.DeclaredOnly | BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);
+                }
+
+                var fieldType = Field.FieldType;
```
Comment (rainersigwald) [inline src/Shared/NodePacketTranslator.cs]: nit: I'd put a blank line after this

**Comment by rainersigwald** [comment]
LGTM

File: src/XMakeCommandLine/XMake.cs [hunk 5]
```diff
@@ -3081,9 +3134,7 @@ private static void ShowHelpMessage()
             Console.WriteLine(AssemblyResources.GetString("HelpMessage_11_LoggerSwitch"));
             Console.WriteLine(AssemblyResources.GetString("HelpMessage_15_ValidateSwitch"));
             Console.WriteLine(AssemblyResources.GetString("HelpMessage_19_IgnoreProjectExtensionsSwitch"));
-#if FEATURE_APPDOMAIN
```
Comment (AndyGerlicher) [inline src/XMakeCommandLine/XMake.cs]: This should still be #if FEATURE_NODE_REUSE, right?

**Comment by AndyGerlicher** [comment]
LGTM

File: src/Shared/NodePacketTranslator.cs (hunk 3 above)
Comment (dsplaisted) [inline src/Shared/NodePacketTranslator.cs]: I don't follow.  The signature of the method allows for the value to be written to the stream or read from it, and this is the implementation that reads from the stream.

**Comment by dsplaisted** [comment]
@dotnet-bot Test this please

File: src/Shared/NodePacketTranslator.cs (hunk 3 above)
Comment (rainersigwald) [inline src/Shared/NodePacketTranslator.cs]: I missed context of being in the read mode. With that, this makes perfect sense.

**Pull request merged**